soh/soh/Enhancements/kaleido.cpp

```diff
@@ -0,0 +1,443 @@
+        if (mTotal >= mRequired && mCount >= mRequired) {
+            totals << '/' << mTotal;
+        }
+        mText = totals.str();
+    }
+
+    void KaleidoEntryIcon::BuildVertices() {
+        int offsetY = 0;
+        int offsetX = 0;
+        // 4 vertices per character, plus one for the preceding icon.
+        Vtx* vertices = (Vtx*)calloc(sizeof(Vtx[4]), mText.length() + 1);
+        // Vertex for the preceding icon.
+        Interface_CreateQuadVertexGroup(vertices, offsetX, offsetY, mIconWidth, mIconHeight, 0);
+        offsetX += 18;
+        for (size_t i = 0; i < mText.length(); i++) {
+            auto charIndex = static_cast<uint16_t>(mText[i] - 32);
+            int charWidth = 0;
+            if (charIndex >= 0) {
+                charWidth = static_cast<int>(Message_GetCharacterWidth(charIndex) * (100.0f / R_TEXT_CHAR_SCALE));
+            }
+            Interface_CreateQuadVertexGroup(&(vertices)[(i + 1) * 4], offsetX, offsetY, charWidth, 16, 0);
+            offsetX += charWidth;
+        }
+        offsetY += FONT_CHAR_TEX_HEIGHT;
+        mWidth = static_cast<int16_t>(offsetX);
+        mHeight = static_cast<int16_t>(offsetY);
+
+        vertices[1].v.ob[0] = 16;
+        vertices[2].v.ob[1] = 16;
+        vertices[3].v.ob[0] = 16;
+        vertices[3].v.ob[1] = 16;
+        vtx = vertices;
+    }
+
+    KaleidoEntryIcon::KaleidoEntryIcon(const char *iconResourceName, int iconFormat, int iconSize, int iconWidth,
+                                       int iconHeight, Color_RGBA8 iconColor, int16_t x, int16_t y, std::string text)
+                                       : mIconResourceName(iconResourceName), mIconFormat(iconFormat), mIconSize(iconSize),
+                                       mIconWidth(iconWidth), mIconHeight(iconHeight), mIconColor(iconColor),
+                                       KaleidoEntry(x, y, std::move(text)) {}
+
+    void KaleidoEntryIcon::RebuildVertices() {
+        free(vtx);
+        vtx = nullptr;
+        BuildVertices();
+    }
+
+    void KaleidoEntryIconCountRequired::Update(PlayState *play) {
+        if (mCount != *mWatch) {
+            mCount = *mWatch;
+            BuildText();
+            RebuildVertices();
+            mAchieved = mCount >= mRequired;
+        }
+
+    }
+
+    KaleidoEntryOcarinaButtons::KaleidoEntryOcarinaButtons(int16_t x, int16_t y) :
+            KaleidoEntryIcon(gItemIconOcarinaOfTimeTex, G_IM_FMT_RGBA, G_IM_SIZ_32b, 32, 32,
+                             Color_RGBA8{ 255, 255, 255, 255 }, x, y, "\x9F\xA5\xA6\xA7\xA8") {
+        CalculateColors();
+        BuildVertices();
+    }
+
+    void KaleidoEntryOcarinaButtons::CalculateColors() {
+        // The commented out code in this function is for handling custom cosmetics for the
+        // A and C buttons. For contrast/accessibility reasons, I've chosen not to respect the
+        // custom cosmetics for now. I've left the code commented out in case that decision gets
+        // reversed later.
```

Comment:
I'd argue if someone needs the cosmetics for accessibility reasons it should be accounted for.

Reply:
I would also like to see them account for cosmetics for accessibility purposes.